src/circuit/gadgets/hashes.rs

```diff
@@ -1,30 +1,101 @@
+use std::fmt::Debug;
+
 use bellperson::{gadgets::num::AllocatedNum, ConstraintSystem, SynthesisError};
 
 use neptune::circuit2::poseidon_hash_allocated as poseidon_hash;
 
 use crate::circuit::gadgets::pointer::{AllocatedPtr, AsAllocatedHashComponents};
 
 use crate::field::LurkField;
-use crate::hash_witness::{ConsName, HashStub, HashWitness, MAX_CONSES_PER_REDUCTION};
+use crate::hash_witness::{ConsName, ConsWitness, ContName, ContWitness, HashName, Stub};
 use crate::store::{HashConst, HashConstants, ScalarPointer, ScalarPtr, Store};
 
-pub struct AllocatedHash<F: LurkField> {
-    preimage: Vec<AllocatedPtr<F>>,
+#[derive(Clone)]
+pub struct AllocatedHash<F: LurkField, PreimageType> {
+    preimage: Vec<PreimageType>,
     digest: AllocatedNum<F>,
 }
 
-pub struct AllocatedHashWitness<'a, F: LurkField> {
+pub type AllocatedPtrHash<F> = AllocatedHash<F, AllocatedPtr<F>>;
+pub type AllocatedNumHash<F> = AllocatedHash<F, AllocatedNum<F>>;
+
+#[derive(Clone, Debug)]
+pub struct Slot<Name: Debug, AllocatedType> {
+    name: Result<Name, ()>,
+    allocated: AllocatedType,
+    consumed: bool,
+}
+
+impl<Name: Debug, F: LurkField, PreimageType> Slot<Name, AllocatedHash<F, PreimageType>> {
+    pub fn new(name: Name, allocated: AllocatedHash<F, PreimageType>) -> Self {
+        Self {
+            name: Ok(name),
+            allocated,
+            consumed: false,
+        }
+    }
+    pub fn new_dummy(allocated: AllocatedHash<F, PreimageType>) -> Self {
+        Self {
+            name: Err(()),
+            allocated,
+            consumed: false,
+        }
+    }
     #[allow(dead_code)]
-    pub(crate) hash_witness: &'a HashWitness<F>, // Sometimes used for debugging.
-    slots: Vec<(Result<ConsName, ()>, AllocatedHash<F>)>,
+    pub fn is_dummy(&self) -> bool {
+        self.name.is_err()
+    }
+    pub fn is_blank(&self) -> bool {
+        self.allocated.digest.get_value().is_none()
+    }
+    pub fn is_consumed(&self) -> bool {
+        self.consumed
+    }
+    fn consume(&mut self) {
+        self.consumed = true
+    }
+}
+
+pub struct AllocatedWitness<'a, VanillaWitness, Name: Debug, AllocatedType> {
+    #[allow(dead_code)]
+    pub(crate) witness: &'a VanillaWitness, // Sometimes used for debugging.
+    slots: Vec<Slot<Name, AllocatedType>>,
+}
+
+impl<'a, VanillaWitness, Name: Debug, F: LurkField, PreimageType>
+    AllocatedWitness<'a, VanillaWitness, Name, AllocatedHash<F, PreimageType>>
+{
+    pub fn assert_final_invariants(&self) {
+        if self.slots[0].is_blank() {
+            return;
+        }
+        let unconsumed = self
+            .slots
+            .iter()
+            .enumerate()
+            .filter(|(_, x)| !x.is_consumed())
+            .map(|(i, x)| (i, &x.name))
+            .collect::<Vec<_>>();
+        assert_eq!(
+            0,
+            unconsumed.len(),
+            "some slots were unconsumed: {:?}",
+            unconsumed
+        );
+    }
```

Comment:
If a hash was allocated when evaluating, it should be consumed by the circuit. Otherwise something is wrong.

Reply:
Actually, this applies even if the hash is a dummy. Without this check, we might allot more slots than needed — and construct dummy hashes that are never used.